src/polkitbackend/polkitbackendactionpool.c

```diff
@@ -454,101 +471,112 @@ ensure_file (PolkitBackendActionPool *pool,
+  g_hash_table_insert (priv->parsed_files, basename, NULL);
+  basename = NULL;
 
  out:
-  g_free (uri);
+  g_free (basename);
+  g_free (path);
 }
 
 static void
 ensure_all_files (PolkitBackendActionPool *pool)
 {
   PolkitBackendActionPoolPrivate *priv;
-  GFileEnumerator *e;
-  GFileInfo *file_info;
-  GError *error;
 
-  priv = polkit_backend_action_pool_get_instance_private (pool);
+  GList *files = NULL;
 
-  e = NULL;
+  priv = polkit_backend_action_pool_get_instance_private (pool);
 
   if (priv->has_loaded_all_files)
     goto out;
 
-  error = NULL;
-  e = g_file_enumerate_children (priv->directory,
-                                 "standard::name",
-                                 G_FILE_QUERY_INFO_NONE,
-                                 NULL,
-                                 &error);
-  if (error != NULL)
+  for (GList *l = priv->directories; l != NULL; l = l->next)
+  {
+    GError *error = NULL;
+    GFileEnumerator *enumerator;
+
+    GFile* file = l->data;
+
+    char *dir_name = g_file_get_path(file);
+
+    enumerator = g_file_enumerate_children (file,
+                                            G_FILE_ATTRIBUTE_STANDARD_NAME,
+                                            G_FILE_QUERY_INFO_NONE,
+                                            NULL,
+                                            &error);
+    if (error != NULL)
     {
-      g_warning ("Error enumerating files: %s", error->message);
-      goto out;
+      g_warning ("Error enumerating files in %s: %s", dir_name, error->message);
     }
-
-  while ((file_info = g_file_enumerator_next_file (e, NULL, &error)) != NULL)
+    else
     {
-      const gchar *name;
-
-      name = g_file_info_get_name (file_info);
-      /* only consider files with the right suffix */
-      if (g_str_has_suffix (name, ".policy"))
-        {
-          GFile *file;
+      GFileInfo *file_info;
+      while ((file_info = g_file_enumerator_next_file (enumerator, NULL, &error)) != NULL)
+      {
+        const gchar *name = g_file_info_get_name (file_info);
+        /* only consider files with the right suffix */
+        if (g_str_has_suffix (name, ".policy"))
+          files = g_list_prepend (files, g_strdup_printf ("%s/%s", dir_name, name));
 
-          file = g_file_get_child (priv->directory, name);
+        g_object_unref (file_info);
+      } /* for all files */
+    }
 
-          ensure_file (pool, file);
+    g_object_unref (enumerator);
+    g_free (dir_name);
+  }
 
-          g_object_unref (file);
-        }
+  /* standard sorting places /etc before /usr as desired */
+  files = g_list_sort (files, (GCompareFunc) g_strcmp0);
 
-      g_object_unref (file_info);
+  for (GList *l = files; l != NULL; l = l->next)
+    {
+      GFile *file = g_file_new_for_path((gchar *)l->data);
+      ensure_file (pool, file);
+      g_object_unref (file);
+    }
 
-    } /* for all files */
+  g_list_free_full (files, g_free);
 
   priv->has_loaded_all_files = TRUE;
 
  out:
```

Comment:
Without the if below, the `goto out` statement from line 527 could be replaced with a simple `return` statement.